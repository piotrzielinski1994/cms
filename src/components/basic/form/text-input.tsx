import FormGroup from './form-group';

type TextInputProps = {};

const TextInput = (props: TextInputProps) => {
  return (
    <FormGroup>
      <input type="text" {...props} />
    </FormGroup>
  );
};

export default TextInput;
