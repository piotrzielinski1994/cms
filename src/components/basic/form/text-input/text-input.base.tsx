import { forwardRef, InputHTMLAttributes } from 'react';
import Form from '../root/form';

type InputProps = Omit<InputHTMLAttributes<HTMLInputElement>, 'value'> & {
  name: string;
  value?: string;
};

const Input = forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  return <input ref={ref} type="text" {...props} value={props.value ?? ''} />;
});

Input.displayName = 'TextInputBase.Input';

const TextInputBase = {
  Root: Form.Group,
  Label: Form.Label,
  Input,
  Error: Form.Error,
};

export default TextInputBase;
