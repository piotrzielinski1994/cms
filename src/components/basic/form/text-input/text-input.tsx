import { cn } from '@/utils/tailwind';
import { DetailedHTMLProps, InputHTMLAttributes } from 'react';
import { Control, FieldValues, Path, useController } from 'react-hook-form';
import Form from '../root/form';

type TextInputProps = Omit<
  DetailedHTMLProps<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>,
  'name'
> & {
  name: string;
  error?: string;
};

type TextInputContainerProps<T extends FieldValues> = Omit<TextInputProps, 'name'> & {
  control: Control<T>;
  name: Path<T>;
};

const classNames = {
  input: ({ isValid }: { isValid: boolean }) => {
    return cn(
      'p-2',
      'border border-solid border-current bg-input',
      'hover:border-foreground/90 hover:ring-foreground/90',
      { '[&:not(:focus)]:text-red-500': !isValid },
      'outline-none ring-inset focus:ring-1 ring-current',
    );
  },
};

const TextInput = ({ error, ...props }: TextInputProps) => {
  return (
    <div className="flex flex-col gap-2">
      <input
        type="text"
        {...props}
        value={props.value ?? ''}
        className={cn(classNames.input({ isValid: !error }), props?.className)}
      />
      <Form.Error>{error}</Form.Error>
    </div>
  );
};

const TextInputContainer = <T extends FieldValues>(props: TextInputContainerProps<T>) => {
  const { control, name, ...rest } = props;
  const { field, fieldState } = useController({ control, name });
  return <TextInput error={fieldState.error?.message} {...rest} {...field} />;
};

export { classNames, TextInput, TextInputContainer };
