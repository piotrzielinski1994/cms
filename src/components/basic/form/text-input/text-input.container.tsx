import { ComponentProps } from 'react';
import { Control, FieldValues, Path, useController } from 'react-hook-form';
import { TextInput } from './text-input';

type TextInputContainerProps<T extends FieldValues> = {
  control: Control<T>;
  name: Path<T>;
} & Omit<ComponentProps<typeof TextInput>, 'name' | 'isValid'>;

const TextInputContainer = <T extends FieldValues>(props: TextInputContainerProps<T>) => {
  const { control, name, ...rest } = props;
  const { field, fieldState } = useController({ control, name });
  return <TextInput {...rest} {...field} error={fieldState.error?.message} />;
};

export { TextInputContainer };
