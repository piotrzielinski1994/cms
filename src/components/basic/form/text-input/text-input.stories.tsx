import Form from '@/components/basic/form/root/form';
import type { Meta, StoryObj } from '@storybook/react';
import { useState, type ComponentProps } from 'react';
import { TextInput } from './text-input';

type Args = ComponentProps<typeof TextInput> & {
  label?: string;
};

const ControlledInput = (props: Args) => {
  const [value, setValue] = useState('');
  return <TextInput {...props} onChange={(e) => setValue(e.currentTarget.value)} value={value} />;
};

const meta: Meta<ComponentProps<typeof ControlledInput>> = {
  component: ControlledInput,
  title: 'Components/TextInput',
  argTypes: {
    label: { control: 'text' },
    error: { control: 'text' },
  },
  args: {
    label: 'Text Input',
  },
};

export const Default: StoryObj<typeof TextInput> = {
  render: ({ label, ...args }: Args) => (
    <>
      <Form.Group>
        <Form.Label htmlFor="input1">{label}</Form.Label>
        <ControlledInput {...args} id="input1" name="input1" />
      </Form.Group>
      <Form.Group>
        <Form.Label htmlFor="input2">Invalid Input</Form.Label>
        <TextInput
          {...args}
          id="input2"
          name="input2"
          value="Invalid value"
          error="Error message"
        />
      </Form.Group>
    </>
  ),
};

export default meta;
