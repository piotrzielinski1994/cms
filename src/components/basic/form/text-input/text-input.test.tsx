import { render } from '@testing-library/react';
import { ComponentProps } from 'react';
import { describe, expect, it, vi } from 'vitest';
import { axe } from 'vitest-axe';
import { TextInput } from './text-input';

describe('TextInput', () => {
  const defaultProps = {
    name: 'name',
    id: 'id',
    label: 'Label',
    value: 'value',
    onChange: vi.fn(),
  } satisfies ComponentProps<typeof TextInput>;

  it('should have no accessibility violations', async () => {
    const { container } = render(<TextInput {...defaultProps} />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const { container } = render(<TextInput {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
