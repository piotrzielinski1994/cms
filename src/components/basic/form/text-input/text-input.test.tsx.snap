// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextInput > should match the snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <input
      class="p-2 border border-solid border-current bg-input hover:border-foreground/90 hover:ring-foreground/90 outline-none ring-inset focus-within:ring-1 ring-current"
      id="id"
      name="name"
      type="text"
      value=""
    />
    <span
      class="min-h-[1em] text-sm text-red-500 leading-none"
      role="alert"
    />
  </div>
</div>
`;
